core/src/main/scala/tamer/Tamer.scala

```diff
@@ -238,87 +226,68 @@ object Tamer {
-        log: LogWriter[Task]
-    ) = log.info(s"stopping producing to $sinkTopic").ignore *>
-      fiber.interrupt *>
-      log.info(s"producer to topic $sinkTopic stopped, running final drain on sink queue").ignore *>
-      sink(kvStream, producer, sinkTopic, keySerializer, valueSerializer, log).run <*
-      log.info("sink queue drained").ignore
-
-    private[tamer] def runLoop(
-        kvChunkQueue: Queue[Chunk[(K, V)]],
-        kvStream: UStream[(K, V)],
-        stateRegistry: ULayer[StateRegistry],
-        sinkRegistry: ULayer[SinkRegistry],
-        log: LogWriter[Task]
-    ) = for {
-      fiber <- sinkStream(kvStream, serdes.keySerializer, serdes.valueSerializer, log).provideSomeLayer[Clock](sinkRegistry)
-      _ <- sourceStream(kvChunkQueue, stateRegistry, log)
-        .ensuringFirst(stopSource(log))
-        .ensuring(
-          drainSink(fiber, kvStream, serdes.keySerializer, serdes.valueSerializer, log)
-            .provideSomeLayer[Clock](sinkRegistry)
-        )
+    private[tamer] def drainSink(fiber: Fiber[Throwable, Unit], stream: UStream[(K, V)], log: LogWriter[Task]) =
+      log.info(s"stopping producing to $sinkTopic").ignore *>
+        fiber.interrupt *>
+        log.info(s"producer to topic $sinkTopic stopped, running final drain on sink queue").ignore *>
+        sink(stream, producer, sinkTopic, keySerializer, valueSerializer, log).run <*
+        log.info("sink queue drained").ignore
+
+    private[tamer] def runLoop(queue: Queue[Chunk[(K, V)]], log: LogWriter[Task]) = for {
+      stream <- ZStream.succeed(ZStream.fromChunkQueueWithShutdown(queue))
+      fiber  <- sinkStream(stream, log)
+      _      <- sourceStream(queue, log).ensuringFirst(stopSource(log)).ensuring(drainSink(fiber, stream, log))
     } yield ()
 
     override val runLoop: ZIO[Clock, TamerError, Unit] = {
       val logic = for {
-        log                  <- logTask
-        _                    <- log.info(s"initializing Tamer with setup: \n$repr")
-        schemaRegistryClient <- Task(new CachedSchemaRegistryClient(config.schemaRegistryUrl, 1000)) // FIXME magic number & manual wiring
-        kvChunkQueue         <- Queue.bounded[Chunk[(K, V)]](config.bufferSize)
-        kvStream             <- UIO(ZStream.fromChunkQueueWithShutdown(kvChunkQueue))
-        stateRegistry        <- UIO(mkRegistry(schemaRegistryClient, stateTopic))
-        sinkRegistry         <- UIO(mkRegistry(schemaRegistryClient, sinkTopic))
-        _                    <- runLoop(kvChunkQueue, kvStream, stateRegistry, sinkRegistry, log).runDrain
+        log   <- logTask
+        _     <- log.info(s"initializing Tamer with setup: \n$repr")
+        queue <- Queue.bounded[Chunk[(K, V)]](config.bufferSize)
+        _     <- runLoop(queue, log).runDrain
       } yield ()
 
-      logic.refineOrDie(tamerErrors)
+      logic.refineOrDie(tamerErrors).provideSomeLayer[Clock](config.schemaRegistryUrl.map(Registry.live(_)).getOrElse(Registry.fake))
```

Comment:
^^ here's where we hide away the creation of the `Registry` so we don't expose it as a requirement to the end user. We create a "live" if we are given a schema registry url, a fake one otherwise. No way for end users to plug their own Confluent-incompatible Registry this way, but maybe for the better